validator/bigquery_table_retention.rego

```diff
@@ -0,0 +1,84 @@
+#
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+package templates.gcp.GCPBigQueryTableRetentionConstraintV1
+
+import data.validator.gcp.lib as lib
+
+deny[{
+	"msg": message,
+	"details": metadata,
+}] {
+	constraint := input.constraint
+	lib.get_constraint_params(constraint, params)
+
+	min_retention_days := lib.get_default(params, "minimum_retention_days", "")
+	max_retention_days := lib.get_default(params, "maximum_retention_days", "")
+
+	asset := input.asset
+	asset.asset_type == "bigquery.googleapis.com/Table"
+
+	# Check if resource is in exempt list
+	exempt_list := params.exemptions
+	matches := {asset.name} & cast_set(exempt_list)
+	count(matches) == 0
+
+	get_diff(asset, min_retention_days, max_retention_days)
+
+	message := sprintf("BigQuery table %v has a retention policy violation.", [asset.name])
```

Comment:
is there a way to surface the violation in the message, eg, x < min or max < x (not necessary if this is overly complicated).

Reply:
I've added a string as an output that will hopefully shed some better light- PTAL!